src/decoder/mod.rs

```diff
@@ -54,7 +64,50 @@ where
     Mp3(mp3::Mp3Decoder<R>),
     #[cfg(feature = "symphonia")]
     Symphonia(symphonia::SymphoniaDecoder),
-    None(::std::marker::PhantomData<R>),
+    None(std::marker::PhantomData<R>),
+}
+
+pub struct DecoderInfo {
+    channels: u16,
+    sample_rate: u32,
+    samples_elapsed: AtomicUsize,
+}
+
+impl DecoderInfo {
+    fn from_source<S>(source: &S) -> Self
+    where
+        S: Source,
+        S::Item: Sample,
+    {
+        Self {
+            channels: source.channels(),
+            sample_rate: source.sample_rate(),
+            samples_elapsed: AtomicUsize::default(),
+        }
+    }
+
+    #[inline]
+    pub fn channels(&self) -> u16 {
+        self.channels
+    }
+
+    #[inline]
+    pub fn sample_rate(&self) -> u32 {
+        self.sample_rate
+    }
+
+    #[inline]
+    pub fn elapsed_samples(&self) -> usize {
+        self.samples_elapsed.load(Ordering::Relaxed)
+    }
+
+    #[inline]
+    pub fn elapsed_duration(&self) -> Option<Duration> {
```

Comment:
This might not be the best name, elapsed sounds like its the time since the song started playing, however it is the position of the song right? So maybe name it .... position, playback_pos or something else